functions.global.php

```diff
@@ -87,4 +87,53 @@ function jetpack_get_migration_data( $option_name ) {
 	$post = get_page_by_title( $option_name, 'OBJECT', 'jetpack_migration' );
 
 	return null !== $post ? maybe_unserialize( $post->post_content_filtered ) : null;
-}
\ No newline at end of file
+}
+
+/**
+ * Intervene upgrade process so Jetpack themes are downloaded with credentials.
+ *
+ * @since 5.3
+ *
+ * @param bool   $preempt Whether to preempt an HTTP request's return value. Default false.
+ * @param array  $r       HTTP request arguments.
+ * @param string $url     The request URL.
+ *
+ * @return array|bool|WP_Error
+ */
+function jetpack_theme_update( $preempt, $r, $url ) {
+	if ( false !== stripos( $url, JETPACK__WPCOM_JSON_API_HOST . '/rest/v1/themes/download' ) ) {
+		$file = $r['filename'];
+		if ( ! $file ) {
+			return new WP_Error( 'problem_creating_theme_file', esc_html__( 'Problem creating file for theme download', 'jetpack' ) );
+		}
+		$theme = pathinfo( parse_url( $url, PHP_URL_PATH ), PATHINFO_FILENAME );
+
+		// Remove filter to avoid endless loop since wpcom_json_api_request_as_blog uses this too.
+		remove_filter( 'pre_http_request', 'jetpack_theme_update' );
```

Comment:
Should this line be at the top of the function before the conditional, or at least before the first return?

Reply:
While it could be placed in line 105, I added it where it is because if update wasn't going to be performed, there's no point in removing it since there won't be a request and the execution flow will simply end returning the WP_Error.
So it's here to remove the filter only if it's necessary and avoid doing it unnecessarily.

Reply:
Okay.  My concern was on a possible bulk action if it may get added multiple times or run on unexpected requests.